### PR TITLE
*-controller-manager: output flags in logical sections

### PR DESCRIPTION
**Pull request by stewart-yu**
image

```
Print *-controller-manager --help flag help in sections.
```

### CODE REVIEW

**Comment by stewart-yu** [comment]
image
no output, seems strange, and the same in master branch, will check
/assign @sttts @luxas @hzxuzhonghu

**Comment by sttts** [comment]
> BTW, no output flags seems strange, should we not let it display?

Yes, would skip those sections.

File: cmd/kube-controller-manager/app/options/options.go [hunk 1]
```diff
@@ -225,38 +225,39 @@ func NewDefaultComponentConfig(insecurePort int32) (componentconfig.KubeControll
...
+	s.Authorization.AddFlags(fss.FlagSet("authorization"))
+
+	s.AttachDetachController.AddFlags(fss.FlagSet("attachdetach controller"))
+	s.CSRSigningController.AddFlags(fss.FlagSet("csrsigning controller"))
+	s.DeploymentController.AddFlags(fss.FlagSet("deployment controller"))
+	s.DaemonSetController.AddFlags(fss.FlagSet("daemonset controller"))
+	s.DeprecatedFlags.AddFlags(fss.FlagSet("deprecated"))
+	s.EndPointController.AddFlags(fss.FlagSet("endpoint controller"))
+	s.GarbageCollectorController.AddFlags(fss.FlagSet("garbagecollector controller"))
+	s.HPAController.AddFlags(fss.FlagSet("horizontalpodautoscaling controller"))
```
Comment (sttts) [inline cmd/kube-controller-manager/app/options/options.go]: maybe CamelCase those strings?

Comment (stewart-yu) [reply cmd/kube-controller-manager/app/options/options.go]: fixed, and add judgement for one option section has no flags

File: cmd/cloud-controller-manager/app/options/options.go [hunk 2]
```diff
@@ -131,24 +131,27 @@ func NewDefaultComponentConfig(insecurePort int32) (componentconfig.CloudControl
...
-	o.Debugging.AddFlags(fs)
-	o.GenericComponent.AddFlags(fs)
-	o.KubeCloudShared.AddFlags(fs)
-	o.ServiceController.AddFlags(fs)
+// Flags returns flags for a specific APIServer by section name
+func (o *CloudControllerManagerOptions) Flags() (fss apiserverflag.NamedFlagSets) {
+	o.CloudProvider.AddFlags(fss.FlagSet("cloudProvider"))
+	o.Debugging.AddFlags(fss.FlagSet("debugging"))
+	o.GenericComponent.AddFlags(fss.FlagSet("generic"))
+	o.KubeCloudShared.AddFlags(fss.FlagSet("kubeCloudShared"))
```
Comment (sttts) [inline cmd/cloud-controller-manager/app/options/options.go]: Disregard my CamlCase comment. Why don't we use human readable section topics? There show up 1:1 in the output. So "Secure serving" instead of "secureServing". And "kubeCloudShared" should be added to "generic" maybe?

Comment (stewart-yu) [reply cmd/cloud-controller-manager/app/options/options.go]: i would like to keep split, but add `generic` also ok, for that two option have no boundary obviouly

Comment (stewart-yu) [reply cmd/cloud-controller-manager/app/options/options.go]: image
:rolling_on_the_floor_laughing:

Comment (stewart-yu) [reply cmd/cloud-controller-manager/app/options/options.go]: When output, thie first lettle would capital, seems ok
image

File: cmd/kube-controller-manager/app/options/options.go [hunk 3]
```diff
@@ -267,7 +268,9 @@ func (s *KubeControllerManagerOptions) AddFlags(fs *pflag.FlagSet, allController
 	var dummy string
 	fs.MarkDeprecated("insecure-experimental-approve-all-kubelet-csrs-for-group", "This flag does nothing.")
 	fs.StringVar(&dummy, "insecure-experimental-approve-all-kubelet-csrs-for-group", "", "This flag does nothing.")
-	utilfeature.DefaultFeatureGate.AddFlag(fs)
+	utilfeature.DefaultFeatureGate.AddFlag(fss.FlagSet("default featuregate"))
```
Comment (sttts) [inline cmd/kube-controller-manager/app/options/options.go]: no "default". That's just code internal. For the user these are "feature flags".

File: cmd/cloud-controller-manager/app/options/options.go [hunk 4]
```diff
@@ -131,24 +131,27 @@ func NewDefaultComponentConfig(insecurePort int32) (componentconfig.CloudControl
...
+	o.Authentication.AddFlags(fss.FlagSet("authentication"))
+	o.Authorization.AddFlags(fss.FlagSet("authorization"))
 
+	fs := fss.FlagSet("misc")
 	fs.StringVar(&o.Master, "master", o.Master, "The address of the Kubernetes API server (overrides any value in kubeconfig).")
 	fs.StringVar(&o.Kubeconfig, "kubeconfig", o.Kubeconfig, "Path to kubeconfig file with authorization and master location information.")
 	fs.DurationVar(&o.NodeStatusUpdateFrequency.Duration, "node-status-update-frequency", o.NodeStatusUpdateFrequency.Duration, "Specifies how often the controller updates nodes' status.")
 
-	utilfeature.DefaultFeatureGate.AddFlag(fs)
+	utilfeature.DefaultFeatureGate.AddFlag(fss.FlagSet("default featuregate"))
```
Comment (sttts) [inline cmd/cloud-controller-manager/app/options/options.go]: as below

**Comment by sttts** [comment]
One nit. Then lgtm.

File: cmd/cloud-controller-manager/app/options/options.go (hunk 4 above)
Comment (stewart-yu) [reply cmd/cloud-controller-manager/app/options/options.go]: done

File: cmd/kube-controller-manager/app/options/options.go [hunk 5]
```diff
@@ -267,7 +268,9 @@ func (s *KubeControllerManagerOptions) AddFlags(fs *pflag.FlagSet, allController
 	var dummy string
 	fs.MarkDeprecated("insecure-experimental-approve-all-kubelet-csrs-for-group", "This flag does nothing.")
 	fs.StringVar(&dummy, "insecure-experimental-approve-all-kubelet-csrs-for-group", "", "This flag does nothing.")
-	utilfeature.DefaultFeatureGate.AddFlag(fs)
+	utilfeature.DefaultFeatureGate.AddFlag(fss.FlagSet("featuregate"))
```
Comment (sttts) [inline cmd/kube-controller-manager/app/options/options.go]: how about just `features`?

Comment (sttts) [reply cmd/kube-controller-manager/app/options/options.go]: There is only one flag inside (`--feature-gates`). This could even go into generic, wdyt?

File: cmd/cloud-controller-manager/app/options/options.go [hunk 6]
```diff
@@ -131,24 +131,27 @@ func NewDefaultComponentConfig(insecurePort int32) (componentconfig.CloudControl
...
-	o.Debugging.AddFlags(fs)
-	o.GenericComponent.AddFlags(fs)
-	o.KubeCloudShared.AddFlags(fs)
-	o.ServiceController.AddFlags(fs)
+// Flags returns flags for a specific APIServer by section name
+func (o *CloudControllerManagerOptions) Flags() (fss apiserverflag.NamedFlagSets) {
+	o.CloudProvider.AddFlags(fss.FlagSet("cloud provider"))
+	o.Debugging.AddFlags(fss.FlagSet("debugging"))
+	o.GenericComponent.AddFlags(fss.FlagSet("generic"))
+	o.KubeCloudShared.AddFlags(fss.FlagSet("generic"))
```
Comment (andrewsykim) [inline cmd/cloud-controller-manager/app/options/options.go]: Wondering if we want to categorize this as "cloud provider" as well. Not blocking though, will think about it and send open another PR if I think it should :)

Comment (stewart-yu) [reply cmd/cloud-controller-manager/app/options/options.go]: which one? `KubeCloudShared`? i think should not:grinning:

**Comment by sttts** [comment]
@deads2k is this approved?
